rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\DowngradePhp70\Rector\Expr;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Expr\ArrayDimFetch;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\PropertyFetch;
+use PhpParser\Node\Expr\StaticCall;
+use PhpParser\Node\Expr\StaticPropertyFetch;
+use Rector\Core\Rector\AbstractRector;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @changelog https://wiki.php.net/rfc/uniform_variable_syntax
+ *
+ * @see \Rector\Tests\DowngradePhp70\Rector\Expr\DowngradeUnnecessarilyParenthesizedExpressionRector\DowngradeUnnecessarilyParenthesizedExpressionRectorTest
+ */
+final class DowngradeUnnecessarilyParenthesizedExpressionRector extends AbstractRector
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Remove parentheses around expressions where they are not necessary thanks to Uniform variable syntax RFC.',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+($f)['foo'];
+($f)->foo;
+($f)->foo();
+($f)::$foo;
+($f)::foo();
+($f)();
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+$f['foo'];
+$f->foo;
+$f->foo();
+$f::$foo;
+$f::foo();
+$f();
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        // TODO: Make PHPStan allow Expr namespace.
+        $dummyClassNameUse = str_contains(Expr::class, '\\');
+        assert($dummyClassNameUse);
+        return [
+            ArrayDimFetch::class,
+            PropertyFetch::class,
+            MethodCall::class,
+            StaticPropertyFetch::class,
+            StaticCall::class,
+            FuncCall::class,
+        ];
+    }
+
+    /**
+     * @param ArrayDimFetch|PropertyFetch|MethodCall|StaticPropertyFetch|StaticCall|FuncCall $node
+     */
+    public function refactor(Node $node): ?Expr
+    {
+        // Re-create the node to strip redundant parentheses around the expression
+        // since PHP 5 cannot deal with them.
+        // In some cases parens are necessary so they will remain even after this rule is applied
+        // but other rules should handle that.
+        return match (true) {
+            $node instanceof ArrayDimFetch => new ArrayDimFetch($node->var, $node->dim),
```

Comment:
Do you think this approach of getting rid of parentheses is reasonable? If yes, I can split it into multiple `NodeRecreator` classes modelled after `NodeNameResolver`. But maybe there is a better way (e.g. clone and clear attributes)?

Reply:
How do you detect if node has parenthesis or not?
Only node that's needs to be changed should be changed.

Reply:
I am not aware of way to detect it since AST abstracts away details like this. Maybe there is some way to get the corresponding concrete syntax? But then there will be nodes where parentheses are necessary so we would still re-create it even though the component that produces the output would put them right back.

Reply:
There might be similar Rector rules that change bracket syntax.

Instead of recrecating nodes, it might be enough to remove original one. It's used to keep the original formatting:

```php
$node->getAttribute(AttributeKey::ORIGINAL_NODE, null);
```

Reply:
That seems to work as well.